join-free sink methods

JoelStienlet opened the pull request:
some small changes to address [issue 75](https://github.com/KjellKod/g3sinks/issues/75)
(avoid joining the thread when passing string arguments to sink methods)
I finally created a new class for a custom-string type that includes its deleter (for remark 3) so that it can be deleted on the same heap as it was allocated.
As this class is common to all sinks, I put it in a new "common" directory.
I tested my "NoJn" methods in my own code, I didn't add the unit tests here yet, because CMakeLists.txt in syslog doesn't seem to include the tests(?), I have to investigate that first.

KjellKod commented on common/g3sinksUtils.hpp:
```diff
@@ -0,0 +1,48 @@
+
+
+#ifndef g3DeepCopyTypes
+#define g3DeepCopyTypes
```
`#pragma once` is a better approach

KjellKod commented on common/g3sinksUtils.hpp:
```diff
@@ -0,0 +1,48 @@
+public:
+    DltrStr() = delete;
+    // the deleter accompanies the memory block to enable it beeing freed in the library (using a different heap, as on Windows):
+    DltrStr(const char * content) {size_t len = strlen(content); _content = std::unique_ptr<char [], std::function<void(char*)>>(new char[len + 1], [](char* b){ delete[](b);} ); memcpy(_content.get(), content, len+1); };
```
Please format the code for improved readability. The compiler doesn't care that it's multiple statements on one line or on multiple lines.

KjellKod commented on logrotate/CMakeLists.txt:
```diff
@@ -15,6 +15,7 @@ SET(DIR_3RDPARTY ${LogRotate_SOURCE_DIR}/../3rdparty)
 
 MESSAGE(" PROJECT_SRC = : ${PROJECT_SRC}")
 INCLUDE_DIRECTORIES(${PROJECT_SRC})
+INCLUDE_DIRECTORIES(${PROJECT_SRC}/../../common)
```
This should not be needed.  If it is then something wonky is happening with CMake. It should be fixed with Cmake instead.

KjellKod commented on logrotate/src/g3sinks/LogRotate.h:
```diff
@@ -14,7 +14,7 @@
 #include <memory>
 
-
+#include "g3sinksUtils.hpp"
```
Please put local includes first in the header file

KjellKod commented on syslog/src/g3log/syslogsink.hpp:
```diff
@@ -57,6 +58,10 @@ class SyslogSink
     void setLevel(LogLevel level, int syslevel);
 
+    // Alternative methods releaving C-string argument lifetime constraints:
+    void setLogHeaderNoJn(g3::DltrStr change) { _header = change.c_str(); }
```
I'd suggest placing each of them right under the original function. That way it's easier to see that there is an alternative function.  The comment is great. I'd have the comment above each function

JoelStienlet commented:
Hello Kjell,
Thank you for your review.

- I'll try to get access to a machine with xcode, to reproduce the compilation errors found by Travis. I don't know yet why he doesn't like my shared_ptr with custom deleter.

- for the `INCLUDE_DIRECTORIES(${PROJECT_SRC}/../../common)` I added this because I added a new "common" directory, so the tree now looks like:
```
|--- common
|         \--- g3sinksUtils.hpp
|--- logrotate
|         |---CMakeLists.txt
|         \--- src
\--- syslog
         |---CMakeLists.txt
         \--- src
```
I modified both CMakeLists.txt to inform cmake to look for header files in "common", and install g3sinksUtils.hpp.
Would you prefer to add a new top-level CMakeLists.txt that would transmit the information about "common" to the two sub-level CMakeLists ?

JoelStienlet commented:
The compilation problem is solved now, I was using a c++17 syntax in the memory allocation.

KjellKod commented:
Yes, a new CMake file that distribute file knowledge in the sub folder makes sense.

KjellKod commented on common/g3sinksUtils.hpp:
```diff
@@ -0,0 +1,57 @@
+        
+    DltrStr(const std::string& content) {
+        size_t len = content.length();
+        _content = std::shared_ptr<char>(new char[len + 1], [](char *b){ delete[](b);} );
```
I'm honestly not sure about this approach. Do you have some references/documentation that will back this up?

KjellKod replied:
L46 - 47

KjellKod replied:
If this is correct (I've to read up on this too) then this class can be greatly simplified.   Please verify this however. https://stackoverflow.com/questions/345003/is-it-safe-to-use-stl-tr1-shared-ptrs-between-modules-exes-and-dlls

JoelStienlet replied:
Interesting read this discussion on Stackoverflow.
Marc Mutz says that a shared_ptr is cross DLL-safe by design, because "there's no such thing as a shared_ptr without a deleter".
So then the deleter could be removed in the class I suggested. Or even remove that class altogether, and pass directly the shared_ptr as argument. But I have some preference for the class, as it allows making changes to the allocation without touching the interface.

But I think there could still be a problem:
let's suppose we have an exe, g3sinks, and another_lib that uses g3sinks.
Let's suppose the exe calls a function from "another_lib" that ends up creating a DltrStr sent to g3sinks. But now suppose that, before the thread from g3log has finished using and deleting that DltrStr, another_lib's dll gets unloaded. Then the deleter inserted into DltrStr (by another_lib) and called in g3sinks' thread would no longer be valid.
This may be a corner case, but I think the documentation should warn about premature unloading of a dll.
All threads from g3sinks must be joined before a lib using g3log gets unloaded.

KjellKod replied:
I wouldn’t jump on what Marc Mutz said without verifying it in the standard. He might be right or wrong. 

There’s a dll loading example in G3log that you can always play with to verify there’s no memory corruption

JoelStienlet replied:
I add another reference:
From Effective C++ Digital Collection: 140 Ways to Improve Your Programming
by Scott Meyers
 "An especially nice feature of tr1::shared_ptr is that it automatically uses its per-pointer deleter to eliminate another potential client error, the "cross-DLL problem."
https://books.google.fr/books?id=U7lTySXdFk0C&lpg=PP1&pg=PT149#v=onepage&q&f=false

KjellKod replied:
Your concern about DltrStr and unloading dll is valid. 
However the recommendation that g3sinks should be joined before dlls are shutting down. I.e G3log is in exit mode after hitting the RAII destruction sequence may be contrary to best practices for G3log. 

It’s assumed that G3log exits (RAII) very late in the shutdown sequence so that any shutdown logging is captured. 

Personally all of this can be avoided by following the recommendation of waiting for the returned future and by not sending a temporary (that expires) to the sink. 
The `call(...) ` was never intended as a high performance frequent calling API but as a thread safe (following the practices we already discussed) way to reach any public function on the sinks API

KjellKod replied:
I guess another way could be to skip the DltStr object but keep the added function to the SyslogSink but having it receive a `shared_ptr` instead. 

The theoretical dll client using the shared_ptr API on the sink is now made aware (or should be) that he shouldn’t shut down before the call is done executing.

KjellKod replied:
For a shutdown sequence the client is always responsible to let future’s etc be done (hard to avoid with a returned future, right?)  so the only thing to worry about is to not sending in a temporary ... which is common thread data passing practice

JoelStienlet replied:
I totally agree that we better test this before adding it to g3sinks. Can I use MinGW for this? (I almost never use Windows, so I've no idea about the right toolchains to use)

JoelStienlet replied:
I'm thinking I should have explained the purpose earlier: the difficulty I had to begin with was for making a python wrapper (to use the same logger in both python and c++):
https://github.com/JoelStienlet/g3logPython
using pybind11.
I could join the threads in the c++ part, but then it blocks the python code until the thread finished, and the python code cannot decide when to join the thread.
Or
I could add these DltStr strings, and then I can join the threads from the python code without worrying for the garbage collector to delete the string in the meantime. (what I observed, and was causing crashes)

KjellKod replied:
I’ll check it out. Very cool usage btw. 

Here’s a dll loading test for nix systems. There’s another one for Windows in the same directory if you need it. 

https://github.com/KjellKod/g3log/blob/master/test_unit/test_linux_dynamic_loaded_sharedlib.cpp

KjellKod replied:
Btw: you AFAIK you don’t need memcopy for the shared_ptr and so you can greatly simplify the code.  ... unless I missed something

JoelStienlet replied:
I'm thinking, perhaps it would be better to manage the lifetime of the strings entirely in the wrapper, and don't change your library at all? I will have to store the allocated strings along the corresponding futures, and "regularly" check if any thread has finished. I don't see any reason why this couldn't be done, so perhaps this is the best way to go after all.

JoelStienlet commented:
After a night's sleep I now think that it's better to avoid this addition to g3sinks. I won't add any "code debt" if I can do otherwise. I'll keep this branch alive, in case it could come in handy to somebody else. Although the "NoJn" functions would definitely simplify my python wrapper, this justification alone is not enough to warrant changes upstream IMO. If other projects would benefit from it too, that would be another story, but for now keeping the interface small is definitely an advantage.
Of course if your opinion is different please let me know, I'd be happy to continue working on this if you feel that it is a valuable addition.
What I'll do anyway though, if you agree, is submit another mini-PR to the readme to explain that argument's lifetime matter, perhaps in a section "technical subtleties" where common pitfalls could be addressed?

KjellKod commented:
I agree. It was a fun exercise and sometimes not adding functionality but working around  feature/limitation is the best approach. 

I would greatly appreciate the documentation update. Please let me know how the workaround goes - will you be using a straight shared_ptr?

JoelStienlet commented:
Hello Kjell,
I started implementing the new solution. My previous approach was passing a deep-copy of the data to the worker thread, who would then be responsible for freeing these data.
The new approach is totally different: the lifetime of the data is now managed through a global data store, that is periodically cleaned-up by an independent thread. When a string is passed from python, a copy is made and added to this data store, and the raw pointer from the string in the store is passed to the worker thread (using the existing interface, with no changes to g3sinks). All data are stored along their respective std::future returned by g3sinks, to keep track of the required lifetime of the data.
The independent clean-up thread periodically checks if any future is == std::future_status::ready, and when ready, the corresponding data in the store are deleted.

KjellKod commented:
Sounds good. I'm looking forward to reviewing it. Is it ready for re-review or are you still working on this?

JoelStienlet commented:
Hello,
Thank you for the discussion and your interest in this development.
I've been able to integrate the new approach entirely in the python wrapper, so that I can avoid making changes to g3sinks upstream (except the include directory for which a did a separate PR, as you have seen).
For me, keeping the project focused, small, and avoiding changes to the API is very valuable, so I'm happy to keep the upstream as-is.

I'm still working on the wrapper (recently added travis and code coverage), the idea is to provide a C++ interface too, so I will add examples using the C++ interface. 
Of course the c++ interface will inherit the convenience of the lifetime-management system required by python.

The python bindings are included entirely in one .cpp, so that it should be possible to compile the "wrapper" without them, for a c++-only project.
Alternatively, it may also provide a good foundation for wrappers for other scripting languages.

When most of the work will be done, and if you're happy with the result, I can transfer the ownership of that wrapper project if you like? Or you can add a link to my wrapper somewhere? My main concern is that people who would need it find it easily, to avoid reinventing the wheel again.
So I'll close this PR now, but we keep in touch, I'll let you know when the wrapper is ready for use "in the field" 😉
